core/network/src/service.rs

```diff
@@ -185,63 +175,89 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> Service<B, S,
 		self.network.lock().average_upload_per_sec()
 	}
 
+	/// Get a clone of the channel to network/libp2p.
+	pub fn network_sender(&self) -> NetworkChan {
+		self.network_sender.clone()
+	}
+
 	/// Called when a new block is imported by the client.
 	pub fn on_block_imported(&self, hash: B::Hash, header: &B::Header) {
-		self.handler.on_block_imported(&mut NetSyncIo::new(&self.network, self.protocol_id), hash, header)
+		let _ = self
+			.protocol_sender
+			.send(ProtocolMsg::BlockImported(hash, header.clone()));
```

Comment:
I wonder if these functions should be changed to take the header by-value explicitly so the clone isn't hidden.